rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake

```diff
@@ -112,53 +103,71 @@ target_include_directories(${rosidl_generate_interfaces_TARGET}${_target_suffix}
   "$<INSTALL_INTERFACE:include>"
 )
 foreach(_pkg_name ${rosidl_generate_interfaces_DEPENDENCY_PACKAGE_NAMES})
+  # Depend on targets generated by this generator in dependency packages
   target_link_libraries(
     ${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE
-    ${${_pkg_name}_TARGETS})
+    ${${_pkg_name}_TARGETS${_target_suffix}})
 endforeach()
+
 target_link_libraries(
   ${rosidl_generate_interfaces_TARGET}${_target_suffix} INTERFACE
   rosidl_runtime_cpp::rosidl_runtime_cpp)
 
+# Make library depend on header generation target
+# TODO(sloretz) make this target name less generic than "__cpp" when other
+# generators no longer depend on it
+add_dependencies(
+  ${rosidl_generate_interfaces_TARGET}${_target_suffix}
+  ${rosidl_generate_interfaces_TARGET}__cpp)
+
+# Make top level generation target depend on this generated library
+add_dependencies(
+  ${rosidl_generate_interfaces_TARGET}
+  ${rosidl_generate_interfaces_TARGET}${_target_suffix})
```

Comment:
When a package calls `rosidl_generate_interfaces(foobar` they're passing in a name to create for a top-level interface generation target. That target name is passed to the individual generators in the variable `${rosidl_generate_interfaces_TARGET}`. Each generator creates their own target to do message generation, and they all do the same pattern where they define a variable called `_target_suffix` and then make their generator specific target with the name `${rosidl_generate_interfaces_TARGET}${_target_suffix}`. The second `add_dependencies()` target is making the top level target depend on the generator specific target. That way, any other target that depends on the top level target will wait until all generators finish before being built. That's part of how targets can use generated messages in the same package.

`rosidl_generator_cpp` only generates headers, so it's library target is a header only library, which in CMake is an `INTERFACE` library. CMake doesn't provide a way for an `INTERFACE` library to directly depend on a file, so there's nothing to make targets who depend on the `INTERFACE` library wait for the header files to be generated. To solve this, `rosidl_generator_cpp` creates an extra custom target called `${rosidl_generate_interfaces_TARGET}__cpp`. CMake does allow custom targets to depend on generated files, and that's all this one does. The first `add_dependency()` makes the generator's `INTERFACE` library depend on that custom target so that anyone who depends on the library will wait to be built until the headers are generated.

Reply:
Added comments explaining these in 8a433e9